### DISCUSSION TITLE
Work Process

### DISCUSSION THREAD

**Discussion by timvahlbrock** [Oeffis/Oeffis#4] [General]
We need to discuss which way of working we want to use and how we want to represent it in GitHub.

**Comment by Paulomart** [comment]
GitHub has some resources on how to work with GitHub in their site documentation. One of them is the [`GitHub Flow`](https://docs.github.com/en/get-started/quickstart/github-flow) it describes how to use branches, pull requests and reviews. I would suggest we use at least this one.

**Comment by Paulomart** [comment]
To breakdown the tasks required to complete the project, we should use milestones. In the milestone we can define overall goals for the iteration. Then we can create issues and assign them to the milestone to further split up the tasks.

**Comment by timvahlbrock** [comment]
## Work Process Proposal

### Work Cycle

Scrum in 2 week cycles

```mermaid
flowchart TD
  biweekly["Bi Weekly (15 min)"]
  refinement["Sprint Relaunch (2h)"]
  biweekly --1 week--> refinement
  refinement --1 week--> biweekly
```

#### Bi Weekly

- 15 mins
- every 2 Weeks
- Discussion of Progress & Problems

#### Sprint Relaunch

- Sprint End (0,5h)
  - escalation of issues that have not been completed
- Sprint review (0,5h)
  - reflection on last sprint
  - to make sure to eradicate problem
- Sprint planning (0,5h)
  - direct assignee only if required, "pull principle" otherwise
  - move most important issues to sprint
  - make sure all the relevant info is there (see Issue Management)
- Backlog Refinement (0,5h)
  - add info for issues in backlog with high priority
  - allows easy & fast issue creation
    - Not all info required on creation
    - Issues can be drawn from backlog even during sprint

#### Roles

- limitation to scrum master
- scrum master is still part of implementation team
- changes after each sprint (at least between persons wanting to do it)

### Issue Mangement

Kanban Board as base

```mermaid
flowchart LR
  Backlog --refinement--> Ready
  Ready --sprint planning--> Planned
  Planned --task started--> Doing
  Doing --implementation done--> Review
  Review --approval--> Done
  Review --denial--> Doing
  START[ ] --issue creation--> Backlog
  START2[ ] --urgent issue creation--> Doing

   style START fill:#FFFFFF00, stroke:#FFFFFF00;
   style START2 fill:#FFFFFF00, stroke:#FFFFFF00;
```

#### Issue Infos
Backlog:
- base info to use in backlog refinement
- urgent issues need to be prepared to go to planned/doing immediately

Ready:
- Priority
- ToDo's
- estimated effort in hours in modified Fibonacci numbers (0.5, 1, 2, 3, 5, 8 hours, etc.)
- maybe test cases (see BDD), edge cases

Planned:
- assignee, if the issue can only be done by a specific person

Doing:
- assignee, if the issue was pulled

Review:
- requested reviewer

Done:
- estimated spent time

### Implementation Style

Usage of Cucumber-Sytle [Behaviour Driven Development](https://cucumber.io/docs/bdd/)

1. Definition of Use Cases, User Stories first in Gherkin, maybe even on the issue
2. implementation of missing actions
3. implementation of productive source code

**Reply by timvahlbrock** [reply]
>Roles
>
>limitation to scrum master

[BDD recommends three roles](https://cucumber.io/docs/bdd/who-does-what/): product owner, tester, developer. Do we want the earlier two as separate roles? Please vote:
💸 Product Owner
🧪 Tester

**Reply by timvahlbrock** [reply]
@Oeffies/core-team

**Reply by JaJako** [reply]
> Do we want the earlier two as separate roles?

In my opinion a separate tester role may be helpful in some cases to separate the different views on one feature (as described by BDD), but is not as necessary as the product owner or developer role.
Although I would have expected to merge tester with developer role (and not product owner)?

> Please vote

~~Meaning 👍or 👎, right?~~ 
💸 Product Owner

**Reply by timvahlbrock** [reply]
I kind of messed up the voting. I wanted everyone that thinks we need a separate product owner to react with 💸 and everyone who thinks we need a separate tester role with 🧪, but I gave no option for "we need neither". But yes, if we ditch the tester role, it would be in the scope of the developer to care for proper testing.

**Reply by JaJako** [reply]
Having an additional issue state "Ready" to separate refined from non-refined issues is definitely a good idea. Haven´t thought that much about how issues will be created but I think it is to be expected that not every issue can be fully specified at the time of creation (or assumptions are changing over time).

Your scrum cycle/process proposed sounds good as it´s not overloaded with many events, but easy to follow without too much organizational overhead. Concerning the frequency of the "Bi Weekly" (or some "Daily Scrum" like meeting) I´m unsure if one meeting per week will be enough. Maybe it should be flexible depending on the time we are able to invest over a week, but once a week at least.

**Reply by timvahlbrock** [reply]
>I´m unsure if one meeting per week will be enough.

Yeah I thought about that too, however it may be difficult to get everyone together two times a week, especially if we want to get stuff done. But one thing that came to my mind, that in my scenario we would only have 1 Daily per Sprint. Options being increasing the length of the sprint or meeting more often. Maybe change sprint duration between semester and holidays? Maybe we need to discuss when we want to do most of the work.

**Comment by JaJako** [comment]
# Scrum Process Proposal
***
```mermaid
flowchart LR
    subgraph init [initial steps]
        direction TB
        projectGoals["☑️ Project Goals"]
        style projectGoals fill:lightgreen
        userStories[User Stories]
        prodBack["(Product) Backlog"]
        projectGoals-->userStories
        userStories-->prodBack
    end
    
    init-->sprintPlanning

    subgraph sprintPlanning [Sprint planning]
       todos["TODOs \n(issues)"]
    end

    sprintPlanning-->sprint

    subgraph sprint ["Sprint (2-3 weeks)"]
        dailyScrum["'Daily' Scrum \n(1-2x/week)"]
    end

    sprint-- Sprint artifacts -->postSprint

    subgraph postSprint [Post-Sprint phase]
        direction TB
        review[Sprint Review]
        retrospective[Sprint Retrospective]  
        review-->retrospective
    end

    postSprint-->sprintPlanning
```

## Roles
* Product Owner
    * someone not actively developing in the project = non-technical point of view
    * feedback on planning & prioritization
    * => by Prof. Conen?
* _Scrum Master_
    * coordination of Scrum Process
    * check compliance with schedule
    * => rotating role? / may be omitted?
* Developer
    * => further breakdown of areas to specialize
        * familiarization with area = more effective work
        * e.g. Frontend & Backend


## Artifacts

### Issues/Tickets
* types
    * Epic/Milestone = see project goals
    * Story/Feature = one specific function (see user stories)
    * Task = small manageable part of Story/Feature (estimated work time 8-16h)
    * Bug = malfunction
* states (simplified transitions)
```mermaid
flowchart LR
    Backlog-->Todo
    subgraph sprint [states within Sprint]
        inProcess[In Process]
        review[Review*]
        Todo-->inProcess
        inProcess-->review
        review-->Done
    end
```
_\* "Review" could also be merged with "In Process" state_
* structure/content
    * What is the given user story? / What is the initial situation (if bug)?
    * What is the goal (of the specific issue)?
    * maybe technical hints
    * estimated time
    * actual time spent (Of whom?)

### Project Goals
* defined already (mandatory, optional)
* => represent as "Epic/Milestone" issues

### User Stories
* What should a user be able to do? (non-technical)
* if important: What does a user have to do technically speaking (which interactions)?
* => represent as "Story/Feature" issues

### Boards
* (Product) Backlog = all "Backlog" issues (separation of "mandatory" and "optional"?)
* Sprint-Board = all Sprint issues with columns for the 3/4 states

### Documentation
* technical: within code / at one specified place within the Repo
* non-technical = user guide: separated from code (within Repo or Wiki)

### Code/App
* buildable, runable Code-Artifact
    * contains all features of the Sprint passed
* branching model: "GitFlow"


## Events

### Sprint-Planning
* choosing Story/Feature issues
* breaking down to manageable tasks (Task issues)

### Sprint
* 2-3 weeks
    * depending on available time of team members
    * starting with 3 weeks?

### "Daily" Scrum
* not daily but 1-2 times a week
    * depending on available time of team members

### Sprint Review
* after each Sprint
* team-internally

### Sprint Retrospective
* after Sprint Review
* together with Prof. Conen (as Product Owner)

## Further

### "Definition of Done"
1. functionality implemented
2. tests implemented (specific coverage?)
3. "good" smell-free code (supporting tools like "SonarQube" in Java?)
4. Documentation (technical and non-technical)
5. Review done (at least with one person, ideally with two)
6. Code comitted to "main" branch

**Reply by timvahlbrock** [reply]
I think our approaches are somewhat similar, nice that you also included ideas on code reviews, code quality and documentation, I totally forgot that. One or two things that came to my mind:

The User Stories Section in "Initial steps": Shouldn't the user stories be defined on the go, iteratively for each feature, so for each issue of milestone?

Epics/Milestones: Definitely a good idea to use those. We might want to think about, what we want to use Milestones for: Features or Deadlines. Milestones would be really nice to group multiple issues together, I think that's not so easy with issues alone and we don't have Epics in GitHub. On the other hand, Milestones would also be great to manage which we Features want to get done to a specific date. We should probably define for both Milestones and Issues, what's used for what.

Product Owner: I thought about, what tasks a product owner might have in our project. You wrote that the product owner has a "non technical point of view" on the project, I would definitely agree with that, but also say that this view is often economical. So tasks would be managing finances, deciding what's in the scope and what isn't. Currently, we don't need to manage money and we decide ourselves what's in the scope and what isn't right? So I guess we don't a product owner? Which would also make my poll obsolete. Thoughts on this? Do you see other tasks, a product owner would adapt?

**Reply by JaJako** [reply]
> Shouldn't the user stories be defined on the go

Maybe I don´t get the difference between initially defining user stories or defining them on the go (apart from time-related questions)? I thought of defining what are the "use cases" based on our project goals (What should a user be able to do?) which would result in some set of planned features.

> Epics/Milestones: Definitely a good idea to use those. We might want to think about, what we want to use Milestones for: Features or Deadlines.

I think my terminology wasn´t ideal. I had Jira´s issue model in mind (Epic, Story, Task, Bug) and haven´t crosschecked if these types are all existing in GitHub, too. Considering milestones on their own, I agree that they would be nice to define a "state of development progress" for one specific date - ideally linked to all issues needed to be done for reaching that state.

> Product Owner: I thought about, what tasks a product owner might have in our project. 

Yes, managing or financial aspects I haven´t had in mind (and agree that we don´t need such). Maybe it`s more of a "Product Owner Lite" role (or even some other name). My main thought was having some person that is not involved in the development process, which would lead to a more "objective" point of view on questions where we as developers are having a more "subjective" point of view after working on something for while.

**Reply by timvahlbrock** [reply]
> I thought of defining what are the "use cases" based on our project goals.

Oh yeah totally works out. We should probably define those at the beginning but also define them more granularly before implementing the actual feature or we would defeat the point of working agile.

**Comment by LaraBiss** [comment]
# Scrum Process Proposal
![Scrum_allgemein](https://github.com/Oeffies/Oeffies/assets/65868918/bb5a1fc3-85a9-421e-811a-cf5c7db564a6)

## Preparation

### Predefined Goals and Tasks

- define overall project goal(s) (done)
- collect use cases
- create UML diagramms (e.g. Use Case Diagramm)
- define milestones for having an overview of time management
- should timetable and current timetable could be helpful

## General

### Roles

- Product Owner: Prof. Conen
- Developer: all
- Scrum Master: one developer

### Weekly

- every 2 weeks (alternate with sprint meeting)
- 15 - 20 minutes
- online via discord
- discuss current state of tasks
- 2 - 3 sentences to what developers have done
- mention problems 
- write short protocol (able to relate to problems when issue takes more time)

## Sprint
![Scrum_Sprint](https://github.com/Oeffies/Oeffies/assets/65868918/dbcdeb65-0c42-4c3a-b066-e1ed3f42e6ad)


### Sprint Review

- 30 minutes
- what issues / tasks were completed during sprint 
- short presentation from every developer
- what issues / tasks were not completed and why
- upadte current timetable to display delay / time ahead

### Sprint Refinement

- 20 minutes
- add tasks / bugs to Backlog
- update priority if necessary

### Sprint Planning

- 40 minutes 
- what issues / tasks for next sprint
- assign / choose who should take the issue / task
- create issues and subdivide in smaller tasks

## Issue Management

### Issue States

- optional: unassigned (open issue)
- todo (assigned for this sprint)
- in process (currently working on this issue)
- review (waiting for beeing reviewed)
- completed (closed)

### Issue Information

- priority
- state
- content
- subtasks
- time total
- time took (current)
- optional: type (task/bugfix)

## Additional Questions

- Where do we log time spend for non-ticket tasks? (e.g. what we have done so far)
- Should we log current total time spend for everyone individually? 
- Is there a good tool for creating and displaying milestones?
- Are there good tools for creating UML diagramms? (Lucid Chart?)
- Should there be a fix or individual number of weeks for every sprint?

**Reply by JaJako** [reply]
> Where do we log time spend for non-ticket tasks?

One option could be not having any tasks without a ticket. We could create some set of "Timetracking Issues" (maybe marked with a special prefix?) which exist for tracking times only. For example there could be an issue named something like "[Time] Meetings". Advantage: Having all logged times together in GitHub (for evaluations). Disadvantage: "Projectmanagement-" and "Development"-issues are mixed (dunno if that's really a problem).

> Should we log current total time spend for everyone individually?

For one specific issue I think it´s not important to have individual current time spend (as separate fields for example), while it should be possible to extract individual times in the end (at least as I understood Prof. Conen). So individual times should be logged, but not necessarily shown own each issue.

> Should there be a fix or individual number of weeks for every sprint?

As far is I read some Scrum guides it is recommended not to change the number of weeks per Sprint too often. But it may be useful for us to adjust the number of weeks if we all having less time (e.g. during examination phase) or if we all having more time than usual.

**Reply by JaJako** [reply]
Explicitly stating the creation of UML diagrams as one step is a good point (dunno if we implicitly had that point in mind already, or not). I agree having some "global" timetable ("roadmap"?) showing should & current state would be very helpful (and milestone as a medium to give an overview of time management).
Writing some kind of protocol if there seems to be no progress (issue takes longer) would be a good idea, indeed (if there is no other visible progress like the commit history).
Concerning issue management I think all three proposals so far are quite similar (apart from some naming or structural details).

**Reply by timvahlbrock** [reply]
>should timetable and current timetable could be helpful

Do you mean something like "when do we want do be done with what"? Should work out nicely with milestones.

**Reply by timvahlbrock** [reply]
> Explicitly stating the creation of UML diagrams as one step is a good point

At what point do you see this? I'm certainly open to having some diagrams, but let's make sure we don't just have have them just "to have diagrams".

**Reply by timvahlbrock** [reply]
> Concerning issue management I think all three proposals so far are quite similar

I think this applies to the proposals in general. I think we should have a provisional process done by tomorrow, so we can get started implementing. We won't get around changing some things as we go anyway, that's what agile is for.

**Comment by timvahlbrock** [comment]
for now closed with solution as set in wiki.
